File: backend/app/domain/schemas/operation.py
from typing import Optional, List, TYPE_CHECKING
from datetime import datetime
from uuid import UUID
from pydantic import BaseModel, ConfigDict
from app.domain.enums import OperationType, OperationStatus
from app.domain.schemas.summaries import ClientSummary, PropertySummary, UserSummary
from app.domain.schemas.user import User

if TYPE_CHECKING:
    from app.domain.schemas.visit import VisitPublic

# History Schema
class OperationStatusHistoryBase(BaseModel):
    operation_id: UUID
    from_status: OperationStatus
    to_status: OperationStatus
    changed_at: datetime
    changed_by_user_id: UUID
    note: Optional[str] = None

class OperationStatusHistoryPublic(OperationStatusHistoryBase):
    id: UUID
    model_config = ConfigDict(from_attributes=True)

# Note Schema for Operations
class OperationNoteBase(BaseModel):
    operation_id: UUID
    text: str

class OperationNoteCreate(BaseModel):
    text: str

class OperationNotePublic(OperationNoteBase):
    id: UUID
    author_user_id: UUID
    author: Optional[UserSummary] = None
    created_at: datetime
    model_config = ConfigDict(from_attributes=True)

# Operation Schemas
class OperationBase(BaseModel):
    type: OperationType
    status: OperationStatus = OperationStatus.INTEREST
    client_id: UUID
    property_id: UUID
    agent_id: UUID

class OperationCreate(OperationBase):
    agent_id: Optional[UUID] = None # Optional in request, set by backend
    note: Optional[str] = None

class OperationUpdate(BaseModel):
    status: Optional[OperationStatus] = None
    note: Optional[str] = None

class OperationPublic(OperationBase):
    id: UUID
    is_active: bool
    created_at: datetime
    updated_at: datetime
    
    # Nested data
    client: Optional[ClientSummary] = None
    property: Optional[PropertySummary] = None
    agent: Optional[User] = None
    status_history: List[OperationStatusHistoryPublic] = []
    notes: List[OperationNotePublic] = []
    visits: List["VisitPublic"] = []
    
    model_config = ConfigDict(from_attributes=True)

# For late evaluation
from app.domain.schemas.visit import VisitPublic
OperationPublic.model_rebuild()

File: frontend/src/types/operation.ts
export enum OperationType {
  SALE = "SALE",
  RENT = "RENT",
}

export enum OperationStatus {
  INTEREST = "INTEREST",
  NEGOTIATION = "NEGOTIATION",
  RESERVED = "RESERVED",
  CLOSED = "CLOSED",
  CANCELLED = "CANCELLED",
}

import { Visit } from "./visit";
import { User } from "./user";

export interface Operation {
  id: string;
  type: OperationType;
  status: OperationStatus;
  client_id: string;
  property_id: string;
  agent_id: string;
  is_active: boolean;
  created_at: string;
  updated_at: string;
  
  // Nested data from backend
  client?: {
    id: string;
    full_name: string;
    phone?: string;
  };
  property?: {
    id: string;
    title: string;
    city: string;
  };
  agent?: User;
  status_history?: OperationStatusHistory[];
  notes?: OperationNote[];
  visits?: Visit[];
}

export interface OperationNote {
  id: string;
  operation_id: string;
  author_user_id: string;
  author?: {
    id: string;
    full_name: string;
  };
  text: string;
  created_at: string;
}

export interface OperationStatusHistory {
  id: string;
  operation_id: string;
  from_status: OperationStatus;
  to_status: OperationStatus;
  changed_at: string;
  changed_by_user_id: string;
  note?: string;
}

export interface OperationCreate {
  type: OperationType;
  status?: OperationStatus;
  client_id: string;
  property_id: string;
  agent_id?: string;
  note?: string;
}

export interface OperationUpdate {
  status?: OperationStatus;
  note?: string;
}
